### PR TITLE
raftstore: add a way to get approximate split keys

### PR DESCRIPTION
**Pull request by Connor1996**
## What have you changed? (mandatory)
This PR adds a way to get approximate split keys. Note that, when `CheckPolicy::APPROXIMATE`, it will call `approximate_split_keys()`, but now split_check_tick will only produce task of `CheckPolicy::SCAN`, I will adjust it in another PR.

## What are the type of the changes? (mandatory)
- New feature (non-breaking change which adds functionality)

## How has this PR been tested? (mandatory)
unit-test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No need

## Does this PR affect tidb-ansible update? (mandatory)
No

### CODE REVIEW

**Comment by Connor1996** [comment]
PTAL @zhangjinpeng1987  @hicqu

File: src/raftstore/store/util.rs [hunk 1]
```diff
@@ -386,58 +433,104 @@ pub fn get_region_approximate_middle_cf(
...
+) -> Result<Vec<Vec<u8>>> {
+    let cf = rocksdb_util::get_cf_handle(db, cfname)?;
+    let start = keys::enc_start_key(region);
+    let end = keys::enc_end_key(region);
+    let range = Range::new(&start, &end);
+    let collection = db.get_properties_of_tables_in_range(cf, &[range])?;
+
+    // cause first element of the iterator will always be returned by step_by(),
+    // for convenience, we insert a useless key in front and then just drop it.
+    let mut keys = vec![b"".to_vec()];
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/raftstore/store/util.rs]: For example, the split size is `3 * PROP_SIZE_INDEX_DISTANCE`
the number stand for the key in `RangeProperties`, `↑` stand for produced split key.

insert: 
start___1___2___3___4___5___6___7....
-↑(drop)------↑----------↑

not insert:
start___1___2___3___4___5___6___7....
-------↑(drop)----↑----------↑

note the start key of region is not included, so if you don't insert a useless key as start key, then  the first splited region's size will be `4 *PROP_SIZE_INDEX_DISTANCE` but not `3 * PROP_SIZE_INDEX_DISTANCE`

File: src/raftstore/store/util.rs [hunk 2]
```diff
@@ -386,58 +433,104 @@ pub fn get_region_approximate_middle_cf(
...
+    }
+    if keys.len() == 1 {
+        return Ok(vec![]);
+    }
+    keys.sort();
+
+    // assume that the size between two keys is PROP_SIZE_INDEX_DISTANCE,
+    // so we make it as split_key every split_size / PROP_SIZE_INDEX_DISTANCE keys.
+    assert!(split_size != 0);
+    let n = (split_size as f64 / PROP_SIZE_INDEX_DISTANCE as f64).ceil() as u64;
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/raftstore/store/util.rs]: wow, great

File: src/raftstore/store/util.rs [hunk 3]
```diff
@@ -386,58 +433,123 @@ pub fn get_region_approximate_middle_cf(
...
+    }
+    keys.sort();
+
+    // assume that the size between two keys is PROP_SIZE_INDEX_DISTANCE,
+    // so we make it as split_key every split_size / PROP_SIZE_INDEX_DISTANCE keys.
+    let len = (keys.len() - 1) as u64;
+    let dist = size as f64 / len as f64;
+    assert!(split_size != 0);
+    let n = (split_size as f64 / dist).ceil() as u64;
+    let mut split_keys = keys
```
Comment (huachaohuang) [inline src/raftstore/store/util.rs]: How about `keys.into_iter().skip(n-1).step_by(n)`?

File: src/raftstore/store/util.rs [hunk 4]
```diff
@@ -386,58 +433,123 @@ pub fn get_region_approximate_middle_cf(
...
+        .step_by(n as usize)
+        .skip(1)
+        .collect::<Vec<Vec<u8>>>();
+
+    if split_keys.len() as u64 > batch_split_limit {
+        split_keys.truncate(batch_split_limit as usize);
+    } else {
+        // make sure not to split when less than max_size for last part
+        let rest = len % n;
+        if rest * PROP_SIZE_INDEX_DISTANCE + split_size < max_size {
```
Comment (huachaohuang) [inline src/raftstore/store/util.rs]: Seems you don't need `PROP_SIZE_INDEX_DISTANCE` anymore.

File: src/raftstore/store/util.rs [hunk 5]
```diff
@@ -386,58 +433,123 @@ pub fn get_region_approximate_middle_cf(
...
+    //         ^(drop)     ^           ^
+    //
+    // note the start key of region is not included, so if we don't insert a
+    // useless key as start key, then the first splited region's size will be
+    // `4 *PROP_SIZE_INDEX_DISTANCE` but not `3 * PROP_SIZE_INDEX_DISTANCE`
+    let mut keys = vec![b"".to_vec()];
+    let mut size = 0;
+    for (_, v) in &*collection {
+        let props = RangeProperties::decode(v.user_collected_properties())?;
+        size += props.get_approximate_keys_in_range(&start, &end);
```
Comment (huachaohuang) [inline src/raftstore/store/util.rs]: You mean `get_approximate_size_in_range()`?

File: src/util/rocksdb/properties.rs [hunk 6]
```diff
@@ -39,8 +39,8 @@ const PROP_ROWS_INDEX_DISTANCE: u64 = 10000;
 const PROP_TOTAL_SIZE: &str = "tikv.total_size";
 const PROP_SIZE_INDEX: &str = "tikv.size_index";
 const PROP_RANGE_INDEX: &str = "tikv.range_index";
-const PROP_SIZE_INDEX_DISTANCE: u64 = 4 * 1024 * 1024;
-const PROP_KEYS_INDEX_DISTANCE: u64 = 40 * 1024;
+pub const PROP_SIZE_INDEX_DISTANCE: u64 = 4 * 1024 * 1024;
```
Comment (huachaohuang) [inline src/util/rocksdb/properties.rs]: Don't need to `pub` anymore.

File: src/raftstore/store/util.rs [hunk 7]
```diff
@@ -386,58 +433,121 @@ pub fn get_region_approximate_middle_cf(
...
+    if keys.len() == 1 {
+        return Ok(vec![]);
+    }
+    keys.sort();
+
+    // use total size of this range and the number of keys in this range to
+    // calculate the average distance between two keys, and we produce a
+    // split_key every `split_size / dist` keys.
+    let len = keys.len();
+    let dist = total_size as f64 / len as f64;
```
Comment (huachaohuang) [inline src/raftstore/store/util.rs]: "dist" is not a common abbreviation, I think "distance" is more clear here.

**Comment by Connor1996** [comment]
PTAL @huachaohuang

**Pull request merged**